test/e2e/ipam/ipamhelper.go

```diff
@@ -0,0 +1,384 @@
+	finder := find.NewFinder(vSphereClient.Client)
+
+	// Find the given folder.
+	folder, err := finder.FolderOrDefault(ctx, folderName)
+	if err != nil {
+		return nil, err
+	}
+
+	// List all VirtualMachines in the folder.
+	managedObjects, err := finder.ManagedObjectListChildren(ctx, folder.InventoryPath+"/...", "VirtualMachine")
+	if err != nil {
+		return nil, err
+	}
+
+	var vm mo.VirtualMachine
+	virtualMachineIPAddresses := map[string]bool{}
+
+	// Iterate over the VirtualMachines, get the `guest.net` property and extract the IP addresses.
+	for _, mobj := range managedObjects {
+		// Get guest.net properties for mobj.
+		if err := vSphereClient.RetrieveOne(ctx, mobj.Object.Reference(), []string{"guest.net"}, &vm); err != nil {
+			return nil, errors.Wrapf(err, "get properties of VM %s", mobj.Object.Reference())
+		}
+		// Iterate over all nics and add IP addresses to virtualMachineIPAddresses.
+		for _, nic := range vm.Guest.Net {
+			if nic.IpConfig == nil {
+				continue
+			}
+			for _, ip := range nic.IpConfig.IpAddress {
+				virtualMachineIPAddresses[ip.IpAddress] = true
+			}
+		}
+	}
+
+	return virtualMachineIPAddresses, nil
+}
+
+func (h *helper) claimIPAddress(ctx context.Context) (_ string, _ *ipamv1.IPAddressClaim, err error) {
+	claim := &ipamv1.IPAddressClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "ipclaim-" + rand.String(32),
+			Namespace: metav1.NamespaceDefault,
+			Labels:    h.labels,
+		},
+		Spec: ipamv1.IPAddressClaimSpec{
+			PoolRef: corev1.TypedLocalObjectReference{
+				APIGroup: ptr.To("ipam.cluster.x-k8s.io"),
+				Kind:     "InClusterIPPool",
+				Name:     "capv-e2e-ippool",
+			},
+		},
+	}
+
+	// Create an IPAddressClaim
+	Byf("Creating IPAddressClaim %s", klog.KObj(claim))
+	if err := h.client.Create(ctx, claim); err != nil {
+		return "", nil, err
+	}
+	// Store claim inside the service so the cleanup function knows what to delete.
+	ip := &ipamv1.IPAddress{}
+
+	var retryError error
+	// Wait for the IPAddressClaim to refer an IPAddress.
+	_ = wait.PollUntilContextTimeout(ctx, time.Second, time.Second*30, true, func(ctx context.Context) (done bool, err error) {
+		if err := h.client.Get(ctx, client.ObjectKeyFromObject(claim), claim); err != nil {
+			retryError = errors.Wrap(err, "getting IPAddressClaim")
+			return false, nil
+		}
+
+		if claim.Status.AddressRef.Name == "" {
+			retryError = errors.Wrap(err, "IPAddressClaim.Status.AddressRef.Name is not set")
+			return false, nil
+		}
+
+		if err := h.client.Get(ctx, client.ObjectKey{Namespace: claim.GetNamespace(), Name: claim.Status.AddressRef.Name}, ip); err != nil {
+			retryError = errors.Wrap(err, "getting IPAddress")
+			return false, nil
+		}
+		if ip.Spec.Address == "" {
+			retryError = errors.Wrap(err, "IPAddress.Spec.Address is not set")
+			return false, nil
+		}
+
+		retryError = nil
+		return true, nil
+	})
+	if retryError != nil {
+		return "", claim, retryError
```

Comment:
```suggestion
		return "", nil, retryError
```

Let's return nil. Otherwise we should document what callers can do with claim in the err != nil case here